app/models/mapping.rb

```diff
@@ -141,17 +134,26 @@ def self.ensure_url(uri)
     end
   end
 
+protected
+
+  def fill_in_scheme
+    self.new_url       = Mapping.ensure_url(new_url)
+    self.suggested_url = Mapping.ensure_url(suggested_url)
+    self.archive_url   = Mapping.ensure_url(archive_url)
+  end
+
   def trim_scheme_host_and_port_from_path
     if path =~ %r{^https?:}
       url = Addressable::URI.parse(path)
       self.path = url.request_uri
     end
   rescue Addressable::URI::InvalidURIError
     # The path isn't parseable, so leave it intact for validations to report
+    Rails.logger.warn("Unparseable URI #{path} in mapping")
   end
 
   def canonicalize_path
-    self.path = site.canonical_path(path) unless (site.nil? || path == '/' || path =~ /^[^\/]/)
+    self.path = site.canonical_path(path) unless site.nil? || path == '/' || path =~ /^[^\/]/
```

Comment:
Probably the thing to do is extract the conditionals to a method, so I'll do that.